File: LeetCode root folder/Count Unhappy Friends.py
# mark and match solution (simulation):61
# ms
# Beats
# 7.98%
class Solution:
    def unhappyFriends(self, n: int, preferences: List[List[int]], pairs: List[List[int]]) -> int:
        hashMap = {}
        for p1, p2 in pairs:
            hashMap[p1] = p2
            hashMap[p2] = p1

        hashSet = set()
        unhappyFriends = set()
        for i in range(n):
            for p in preferences[i]:
                if p == hashMap[i]:
                    break
                preferredPair = (min(i, p), max(i, p))
                if preferredPair in hashSet:
                    # print(preferredPair)
                    unhappyFriends.add(i)
                    unhappyFriends.add(p)
                else:
                    hashSet.add(preferredPair)
        return len(unhappyFriends)

File: LeetCode root folder/Count Unhappy Friends.cpp
//simulation solution:75
//ms
//Beats
//14.69%

class Solution {
  public:
    int unhappyFriends(int n, vector<vector<int>>& preferences, vector<vector<int>>& pairs) {
        unordered_map<int, int> hashMap;
        for (vector<int> e : pairs) {
          int p1 = e[0];
          int p2 = e[1];
          hashMap[p1] = p2;
          hashMap[p2] = p1;
        }

        unordered_set<string> hashSet;
        unordered_set<int> unhappyFriends;
        for (int i = 0; i < n; i++) {
            for (int p : preferences[i]) {
                if (p == hashMap[i]) {
                    break;
                }
                string preferredPair = to_string(min(i, p)) + ',' + to_string(max(i, p));
                if (hashSet.find(preferredPair) != hashSet.end()) {
                    unhappyFriends.insert(i);
                    unhappyFriends.insert(p);
                }
                else {
                    hashSet.insert(preferredPair);
                }
            }
        }
        return unhappyFriends.size();
    }
};